002-es6-modules.md

```diff
@@ -0,0 +1,274 @@
+    - Represents a read-only snapshot of a module's exports.
+    
+### Operations
+    
+* [ParseModule](https://tc39.github.io/ecma262/#sec-parsemodule)
+    - Creates a SourceTextModuleRecord from source code.
+    
+* [HostResolveImportedModule](https://tc39.github.io/ecma262/#sec-hostresolveimportedmodule)
+    - A hook for when an import is exactly performed.
+    
+* [CreateImportBinding](https://tc39.github.io/ecma262/#sec-createimportbinding)
+    - A means to create a shared binding (variable) from an export to an import. Required for the "live" binding of imports.
+
+* [ModuleNamespaceCreate](https://tc39.github.io/ecma262/#sec-modulenamespacecreate)
+    - Provides a means of creating a list of exports manually, used so that CommonJS `module.exports` can create ModuleRecords that are prepopulated.
+
+---
+
+### [WhatWG Loader](https://github.com/whatwg/loader)
+
+---
+
+Discusses the design of module metadata in a [Registry](https://whatwg.github.io/loader/#registry). All actions regarding the Registry are synchronous.
+
+**NOTE** It is not Node's intent to implement the asynchronous pipeline in the Loader specification. There is discussion about including a synchronous pipeline in the specification as an addendum.
+
+
+### [Summary Video on Youtube](youtube.com/watch?v=NdOKO-6Ty7k)
+
+## Additional Structures Required
+
+### DynamicModuleRecord
+
+A Module Record that presents a view of an Object for its `[[Namespace]]` rather than coming from an environment record.
+
+The list of exports is frozen upon construction. No new properties may be added. No properties may be removed.
+
+## Algorithm
+
+When `require()`ing a file.
+
+1. Determine if file is ES6 or CommonJS (CJS).
+2. If CJS
+	1. Evaluate immediately
+	2. Produce a DynamicModuleRecord from `module.exports`
+3. If ES6
+	1. Parse for `import`/`export`s and keep record, in order to create bindings
+	2. Gather all submodules by performing `require` recursively
+		* See circular dep semantics below 
+	3. Connect `import` bindings for all relevant submodules (see [ModuleDeclarationInstantiation](https://tc39.github.io/ecma262/#sec-moduledeclarationinstantiation))
+	4. Evaluate
+
+## Semantics
+
+### Determining if source is an ES6 Module
+
+If a module can be parsed outside of the ES6 Module goal, it will be treated as a [Script](https://tc39.github.io/ecma262/#sec-parse-script). Otherwise it will be parsed as [ES6](https://tc39.github.io/ecma262/#sec-parsemodule).
+
+In pseudocode, it looks somewhat like:
+
+```
+try {
+  v8::Script::Compile
+}
+catch (e) {
+  v8::Script::CompileModule
+}
+```
+
+V8 may or may not choose to use parser fallback to combine this into one step.
+
+### CommonJS consuming ES6
+
+#### default exports
+
+ES6 modules only ever declare named exports. A default export just exports a property named `default`.
+
+Given
+
+```javascript
+let foo = 'my-default';
+export default foo;
+```
+
+```javascript
+require('./es6');
```

Comment:
> Calling my ideas "delusions" is a quick way to turn this conversation ugly.

I say that because it appears that you are stating something which i do not see, nor believe to be true.

> Please stop stating this. This is not at all true. They are different parsing goals. You can make them compatible, but they are not at all a subset. There are obvious things you can do in each that you cannot do in the other.

What can you do in an ES6 that is currently unable to be transpiled?

> This is the danger of basing your work on an incomplete and un-implemented spec.

We are still capable of all this proposal under ES6 ( http://www.ecma-international.org/ecma-262/6.0/index.html )

> There are aspects that are not yet finalized, but are definitely coming, and top-level await is one of them.

Any public signals?

> Browser work has been done carefully to allow this definitely-coming possibility. Node should proceed in the same way.

Node has a slightly different situation, we have a very large existing number of modules. We have well defined module semantics. Browser is greenfield with globals being the way to support existing libraries.

> What does "run to completion" mean to you?

* https://en.wikipedia.org/wiki/Run_to_completion_scheduling

My understanding was that upon [ScriptEvaluation](https://tc39.github.io/ecma262/#sec-runtime-semantics-scriptevaluation) or [ModuleEvaluation](https://tc39.github.io/ecma262/#sec-moduleevaluation) you were guaranteed that a completion record for that operation would be generated as is listed at the end of both of those operations, or the operation would be stuck in a loop. Top level await can be forever stuck, without causing the operation to block.